packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts

```diff
@@ -0,0 +1,191 @@
+		Object.entries(alias)
+			.map(([key, value]) => {
+				let resolvedAliasPath;
+				try {
+					resolvedAliasPath = require.resolve(value);
+				} catch (e) {
+					// this is an alias for package that is not installed in the current app => ignore
+					resolvedAliasPath = "";
+				}
+
+				return [key, resolvedAliasPath.replace(/\.cjs$/, ".mjs")];
+			})
+			.filter((entry) => entry[1] !== "")
+	);
+	const UNENV_ALIAS_RE = new RegExp(
+		`^(${Object.keys(aliasAbsolute).join("|")})$`
+	);
+
+	build.onResolve({ filter: UNENV_ALIAS_RE }, (args) => {
+		// Resolve the alias to its absolute path and potentially mark it as external
+		return {
+			path: aliasAbsolute[args.path],
+			external: external.includes(alias[args.path]),
+		};
+	});
+}
+
+/**
+ * Inject node globals defined in unenv's `inject` config via virtual modules
+ */
+function handleNodeJSGlobals(
+	build: PluginBuild,
+	inject: Record<string, string | string[]>
+) {
+	const UNENV_GLOBALS_RE = /_virtual_unenv_global_polyfill-([^.]+)\.js$/;
+
+	build.initialOptions.inject = [
+		...(build.initialOptions.inject ?? []),
+		//convert unenv's inject keys to absolute specifiers of custom virtual modules that will be provided via a custom onLoad
+		...Object.keys(inject).map((globalName) =>
+			nodePath.resolve(
+				getBasePath(),
+				`_virtual_unenv_global_polyfill-${globalName}.js`
+			)
+		),
+	];
+
+	build.onResolve({ filter: UNENV_GLOBALS_RE }, ({ path }) => ({ path }));
+
+	build.onLoad({ filter: UNENV_GLOBALS_RE }, ({ path }) => {
+		// eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
+		const globalName = path.match(UNENV_GLOBALS_RE)![1];
+		const globalMapping = inject[globalName];
+
+		if (typeof globalMapping === "string") {
+			const globalPolyfillSpecifier = globalMapping;
+
+			return {
+				contents: `
+								import globalVar from "${globalPolyfillSpecifier}";
+
+								${
+									/*
+								// ESBuild's inject doesn't actually touch globalThis, so let's do it ourselves
+								// by creating an exportable so that we can preserve the globalThis assignment if
+								// the ${globalName} was found in the app, or tree-shake it, if it wasn't
+								// see https://esbuild.github.io/api/#inject
+								*/ ""
+								}
+								const exportable =
+									${
+										/*
+									// mark this as a PURE call so it can be ignored and tree-shaken by ESBuild,
+									// when we don't detect 'process', 'global.process', or 'globalThis.process'
+									// in the app code
+									// see https://esbuild.github.io/api/#tree-shaking-and-side-effects
+									*/ ""
+									}
+									/* @__PURE__ */ (() => {
+										${
+											/*
+										// TODO: should we try to preserve globalThis.${globalName} if it exists?
+										*/ ""
+										}
+										return globalThis.${globalName} = globalVar;
+									})();
+
+								export {
+									exportable as '${globalName}',
+									exportable as 'globalThis.${globalName}',
```

Comment:
Does this work?

Reply:
Our fixture doesn't test it but it is being used in the node.js compat matrix app that Igor and James C built. I believe that gives us enough confidence for today. But I'll see if I can add a quick check in the fixture.

Reply:
Added 0ec2e8632ced68a2db402223a73f84b893643be7